staging/src/k8s.io/apimachinery/pkg/util/proxy/stream-translator.go

```diff
@@ -0,0 +1,132 @@
+type StreamTranslatorHandler struct {
+	// Location is the location of the upstream proxy. It is used as the location to Dial on the upstream server
+	// for upgrade requests.
+	Location *url.URL
+	// Transport provides an optional round tripper to use to proxy. If nil, the default proxy transport is used
+	Transport http.RoundTripper
+	// FlushInterval controls how often the standard HTTP proxy will flush content from the upstream.
+	FlushInterval time.Duration
+	// MaxBytesPerSec controls the maximum rate for an upstream connection. No rate is imposed if the value is zero.
+	MaxBytesPerSec int64
+	// Responder is passed errors that occur while setting up proxying.
+	Responder ErrorResponder
+}
+
+// NewStreamTranslatorHandler creates a new proxy handler with a default flush interval. Responder is required for returning
+// errors to the caller.
+func NewStreamTranslatorHandler(location *url.URL, transport http.RoundTripper, responder ErrorResponder) *StreamTranslatorHandler {
+	return &StreamTranslatorHandler{
+		Location:      normalizeLocation(location),
+		Transport:     transport,
+		FlushInterval: defaultFlushInterval,
+		Responder:     responder,
+	}
+}
+
+func (h *StreamTranslatorHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// Create WebSocket server, including particular streams requested.
+	streamOpts, _ := streamOptionsFromRequest(req)
+	ctx, ok := webSocketServerStreams(req, w, streamOpts, defaultIdleConnectionTimeout)
+	if !ok {
+		err := fmt.Errorf("Unable to create websocket streams")
+		klog.Infof("websocket createStreams error: %v", err)
+		h.Responder.Error(w, req, err)
+		return
+	}
+	defer ctx.conn.Close()
+
+	// Create the SPDY client connection proxied through kubelet.
+	tlsConfig, err := utilnet.TLSClientConfig(h.Transport)
+	if err == nil {
+		// Manually set NextProtos to "http/1.1", since http/2.0 will NOT upgrade a connection.
+		tlsConfig.NextProtos = []string{"http/1.1"}
+	} else {
+		klog.Infof("Unable to unwrap transport %T to get at TLS config: %v", h.Transport, err)
+	}
+	spdyRoundtripper := spdy.NewRoundTripper(tlsConfig)
+	spdyExecutor, err := spdy2.NewSPDYExecutorForTransports(spdyRoundtripper, spdyRoundtripper, "POST", h.Location)
+	if err != nil {
+		klog.Infof("create SPDY executor error: %v", err)
+		h.Responder.Error(w, req, err)
+		return
+	}
+
+	// Wire the WebSocket server stream output to the SPDY client input.
+	opts := spdy2.StreamOptions{}
+	if streamOpts.Stdin {
+		opts.Stdin = ctx.stdinStream
+	}
+	if streamOpts.Stdout {
+		opts.Stdout = ctx.stdoutStream
+	}
+	if streamOpts.Stderr {
+		opts.Stderr = ctx.stderrStream
+	}
+	if streamOpts.TTY {
+		opts.Tty = true
+		opts.TerminalSizeQueue = &translatorSizeQueue{resizeChan: ctx.resizeChan}
+	}
+	// Start the SPDY client with connected streams.
+	err = spdyExecutor.Stream(opts)
+	if err != nil {
+		klog.Infof("spdyExecutor Stream() error: %v", err)
+		h.Responder.Error(w, req, err)
```

Comment:
I investigated ` [sig-node] PrivilegedPod [NodeConformance] should enable privileged commands [LinuxOnly]` failing test. 

Test expects an error(non zero exit code should be passed from api server to kubectl) when command is executed for non-privileged container but it seems that `h.Responder.Error` sends unrecognized error message rather than `apierrors.StatusError` type.